File: modules/x-engine-native-direct/iOS/Class/Native_direct.m
//
//  Native_direct.m
//  direct
//
//  Created by zk on 2020/9/7.
//  Copyright © 2020 edz. All rights reserved.


#import "Native_direct.h"
#import "NativeContext.h"
#import "iDirectManager.h"
#import "iDirect.h"
@interface Native_direct()
@property (nonatomic, strong) NSMutableDictionary<NSString*, id<iDirect>> * directors;
@end

@implementation Native_direct
NATIVE_MODULE(Native_direct)

 - (NSString*) moduleId{
    return @"com.zkty.native.direct";
}

- (int) order{
    return 0;
}

- (instancetype)init
{
    self = [super init];
    self.directors=[NSMutableDictionary new];
    return self;
}

- (void)afterAllNativeModuleInited{
   NSArray* modules= [[NativeContext sharedInstance]  getModulesByProtocol:@protocol(iDirect)];
    for(id<iDirect> direct in modules){
        [self.directors setObject:direct forKey:[direct scheme]];
    }
}

- (void)back: (NSString*) scheme host:(NSString*) host pathname:(NSString*) pathname{
    id<iDirect> direct = [self.directors objectForKey:scheme];
    [direct back:host pathname:pathname];
}

- (void)push: (NSString*) scheme
        host:(nullable NSString*) host
        pathname:(NSString*) pathname
        query:(nullable NSDictionary<NSString*,NSString*>*) query
        params:(NSDictionary<NSString*,NSString*>*) params {
    id<iDirect> direct = [self.directors objectForKey:scheme];
    [direct push:[direct protocol] host:host pathname:pathname query:query params:params];
}

@end
 

File: modules/x-engine-native-direct/android/native-direct/src/main/java/com/zkty/nativ/direct/NativeDirect.java
package com.zkty.nativ.direct;


import android.text.TextUtils;

import com.zkty.nativ.core.NativeContext;
import com.zkty.nativ.core.NativeModule;
import com.zkty.nativ.jsi.webview.XEngineWebView;
import com.zkty.nativ.jsi.webview.XWebViewPool;

import java.util.HashMap;
import java.util.List;
import java.util.Map;


public class NativeDirect extends NativeModule implements IDirectManager {
    private static int j = 1;


    {
        directors = new HashMap<>();

    }

    private Map<String, IDirect> directors;

    @Override
    public String moduleId() {
        return "com.zkty.native.direct";
    }

    @Override
    public int order() {
        return 0;
    }

    @Override
    public void afterAllNativeModuleInited() {
        List<NativeModule> modules = NativeContext.sharedInstance().getModulesByProtocol(IDirect.class);
        for (NativeModule iDirect : modules) {
            if (iDirect instanceof IDirect) {
                directors.put(((IDirect) iDirect).scheme(), (IDirect) iDirect);
            }
        }
    }

    @Override
    public void push(String scheme, String host, String pathname, String fragment, Map<String, String> query, Map<String, String> params) {
        if (TextUtils.isEmpty(host) && XWebViewPool.sharedInstance().getCurrentWebView() != null) {
            XEngineWebView xEngineWebView = XWebViewPool.sharedInstance().getCurrentWebView();
            host = xEngineWebView.getHistoryModel().host;
            pathname = xEngineWebView.getHistoryModel().pathname;
        }

        IDirect iDirect = directors.get(scheme);
        iDirect.push(iDirect.protocol(), host, pathname, fragment, query, params);
    }

    @Override
    public void back(String scheme, String host, String fragment) {
        IDirect iDirect = directors.get(scheme);
        iDirect.back(host, fragment);
    }
}
